packages/block-library/src/post-content/index.php

```diff
@@ -37,3 +37,21 @@ function register_block_core_post_content() {
 	);
 }
 add_action( 'init', 'register_block_core_post_content' );
+
+/**
+ * Only show the `core/post-content` block when editing templates and template parts.
+ * This is to prevent infinite recursions (post content containing a Post Content block
+ * that attempts to embed the post content it is part of).
+ *
+ * @param bool|array $allowed_block_types Array of block type slugs, or
+ *                                        boolean to enable/disable all.
+ * @param WP_Post    $post                The post resource data.
+ */
+function disallow_core_post_content_block_outside_templates( $allowed_block_types, $post ) {
+	if ( in_array( $post->post_type, array( 'wp_template', 'wp_template_part' ), true ) ) {
+		return $allowed_block_types;
+	}
+	return array_diff( $allowed_block_types, array( 'core/post-content' ) );
```

Comment:
I think this doesn't work because the `$allowed_block_types` arg that's passed to this filter isn't an array but `true` (meaning "enable all blocks") :cry:

Reply:
For the time being, we'll probably address this on the client side, see e.g. https://developer.wordpress.org/block-editor/developers/filters/block-filters/#using-a-deny-list.

For the long term, I've filed https://github.com/WordPress/wordpress-develop/pull/419.